catboost/private/libs/algo/leafwise_scoring.cpp

```diff
@@ -341,22 +342,39 @@ static void CalcScoresForSubCandidate(
             updateSplitScoreClosure);
     };
 
-    // TODO(ilyzhin) make caching for nonsymmetric trees
-    if (!ctx->UseTreeLevelCaching() || ctx->Params.ObliviousTreeOptions->GrowPolicy != EGrowPolicy::SymmetricTree) {
-        TVector<TBucketStats> stats;
-        stats.yresize(bucketCount);
-
-        for (auto leaf : leafs) {
-            const auto leafBounds = fold.LeavesBounds[leaf];
-            if (leafBounds.Empty()) {
-                continue;
-            }
+    TArrayRef<TBucketStats> statsRef = statsForSubtractionTrick.GetStatsRef();
+    TArrayRef<TBucketStats> parentStatsRef = statsForSubtractionTrick.GetParentStatsRef();
+    TArrayRef<TBucketStats> siblingStatsRef = statsForSubtractionTrick.GetSiblingStatsRef();
 
-            extractBucketIndex(leafBounds);
-            for (int dim : xrange(approxDimension)) {
-                calcStats(leafBounds, dim, stats);
+    auto calcStatsScores = [&] (TArrayRef<TBucketStats> stats) {
```

Comment:
Please update the canonical test outputs for depth-wise trees in pytest and python-package

Reply:
done